pytorch_lightning/trainer/trainer.py

```diff
@@ -776,10 +776,20 @@ def fit(
         # download the data and do whatever transforms we need
         # do before any spawn calls so that the model can assign properties
         # only on proc 0 because no spawn has happened yet
-        if not self._is_data_prepared:
+        # for slurm or torchelastic, prepare the data once per node
+        if (
+            not (self.use_ddp or self.use_ddp2)
+            or (self.is_slurm_managing_tasks and int(os.environ["SLURM_LOCALID"]) == 0)
+            # torchelastic or general non_slurm ddp
+            or (
+                "WORLD_SIZE" in os.environ
+                and ("GROUP_RANK" in os.environ or "NODE_RANK" in os.environ)
+                and int(os.environ["LOCAL_RANK"]) == 0
+            )
+        ):
```

Comment:
I have two concerns in terms of code quality:
1. these extra 10 lines of code make the fit method even bigger than it already should be.
2. it is hard to read (and to debug). suggestion: break it down like this:
```python 
condition1 = ...
condition2 = ... 
... 

if condition1 and condition2 and not condition3 ...
```

and choose good names for these variables. how does this sound?